File: services/tenantd/csharp/src/CtlFlow.Tenancy.Tenantd.Service/Aggregation/Documents/WorkspaceDocument.cs
namespace CtlFlow.Tenancy.Tenantd.Service.Aggregation.Documents;

internal sealed class WorkspaceDocument
{
    public required string ApiVersion { get; init; }

    public required string Kind { get; init; }

    public required ObjectMetaDocument Metadata { get; init; }

    public required WorkspaceSpecDocument Spec { get; init; }

    public ResourceStatusDocument? Status { get; init; }
}

internal sealed class WorkspaceSpecDocument
{
    public required string TenantId { get; init; }

    public required string DisplayName { get; init; }

    public required string WorkspaceAddress { get; init; }

    public required InitialMembershipDocument[] InitialMemberships
    {
        get;
        init;
    }

    public required BaselinePackageDocument[] BaselinePackages { get; init; }
}

internal sealed class WorkspaceListDocument
{
    public required string ApiVersion { get; init; }

    public required string Kind { get; init; }

    public required ListMetaDocument Metadata { get; init; }

    public required WorkspaceDocument[] Items { get; init; }
}

File: services/tenantd/tests/support/workspace-document.ts
export interface WorkspaceDocument {
  readonly apiVersion: "tenancy.ctlflow.com/v1alpha1";
  readonly kind: "Workspace";
  readonly metadata: {
    readonly name: string;
    readonly resourceVersion: string;
    readonly creationTimestamp: string;
  };
  readonly spec: {
    readonly tenantId: string;
    readonly displayName: string;
    readonly workspaceAddress: string;
    readonly initialMemberships: readonly {
      readonly userId: string;
      readonly standing: string;
    }[];
    readonly baselinePackages: readonly {
      readonly packageId: string;
      readonly packageVersion: string;
    }[];
  };
  readonly status: {
    readonly lifecycle: string;
    readonly revision: number;
    readonly provisioningGeneration: number;
    readonly currentOperation?: {
      readonly id: string;
      readonly kind: string;
    };
    readonly conditions: readonly {
      readonly owner: string;
      readonly state: string;
      readonly ownerRevision?: number;
      readonly reason?: string;
      readonly lastTransitionTime: string;
    }[];
  };
}
